File: Backend/services/WebSocket.py
import asyncio
import websockets
import json
import logging
from datetime import datetime, timezone
from typing import Set, Dict
import os
import sys
from pathlib import Path
from redis.asyncio import Redis
from dotenv import load_dotenv

# --- Logging ---
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("websocket-streamer")

# --- ENV Setup ---
sys.path.append(str(Path(__file__).resolve().parent))
if not os.environ.get("ENV"):
    load_dotenv(Path(__file__).resolve().parents[1] / ".env")

FINNHUB_API_KEY = os.environ["FINNHUB_API_KEY"]
REDIS_URL = os.environ["REDIS_URL"]
FINNHUB_WS_URL = f"wss://ws.finnhub.io?token={FINNHUB_API_KEY}"

SYMBOLS_KEY = "stock:symbols"
PRICE_PREFIX = "stock:price:"
TRADE_PREFIX = "stock:trade:"
OHLCV_PREFIX = "stock:ohlcv:"

# --- In-memory OHLCV Buffers ---
ohlcv_buffer: Dict[str, list] = {}

# --- Redis Utils ---
async def get_symbols(redis: Redis) -> Set[str]:
    try:
        symbols = await redis.smembers(SYMBOLS_KEY)
        return {s.decode() if isinstance(s, bytes) else s for s in symbols}
    except Exception as e:
        logger.error(f"[Redis] Failed to fetch symbols: {e}")
        return set()

async def flush_trade(redis: Redis, trade: dict):
    symbol = trade.get("s")
    price = trade.get("p")
    timestamp = trade.get("t")
    volume = trade.get("v", 0)
    now = datetime.now(timezone.utc).isoformat()

    if not symbol or price is None:
        return

    trade_info = {
        "price": price,
        "timestamp": timestamp,
        "volume": volume,
        "updated_at": now
    }

    try:
        pipe = redis.pipeline()
        pipe.set(f"{PRICE_PREFIX}{symbol}", price)
        pipe.hset(f"{TRADE_PREFIX}{symbol}", mapping=trade_info)
        await pipe.execute()
    except Exception as e:
        logger.error(f"[Redis] Pipeline failed for trade: {e}")

    # Update OHLCV buffer
    if symbol not in ohlcv_buffer:
        ohlcv_buffer[symbol] = []
    ohlcv_buffer[symbol].append({"price": price, "volume": volume})

async def flush_ohlcv(redis: Redis):
    now = datetime.now(timezone.utc).isoformat()
    for symbol, trades in ohlcv_buffer.items():
        if not trades:
            continue
        prices = [t["price"] for t in trades]
        volumes = [t["volume"] for t in trades]

        open_ = prices[0]
        high = max(prices)
        low = min(prices)
        close = prices[-1]
        volume = sum(volumes)

        ohlcv = {
            "open": open_,
            "high": high,
            "low": low,
            "close": close,
            "volume": volume,
            "updated_at": now
        }

        try:
            await redis.hset(f"{OHLCV_PREFIX}{symbol}", mapping=ohlcv)
        except Exception as e:
            logger.error(f"[Redis] Failed to flush OHLCV for {symbol}: {e}")

    ohlcv_buffer.clear()

# --- Streamer ---
async def stream_trades():
    redis = Redis.from_url(REDIS_URL, decode_responses=True)
    logger.info("[Redis] Connected ✅")

    reconnect_delay = 3
    max_delay = 60

    async def periodic_ohlcv_flush():
        while True:
            await asyncio.sleep(10)
            await flush_ohlcv(redis)

    asyncio.create_task(periodic_ohlcv_flush())

    while True:
        try:
            logger.info("[WS] Connecting to Finnhub...")
            async with websockets.connect(FINNHUB_WS_URL, ping_interval=20, ping_timeout=10) as ws:
                logger.info("[WS] Connected ✅")
                reconnect_delay = 3

                symbols = await get_symbols(redis)
                if not symbols:
                    logger.warning("[WS] No symbols to subscribe — sleeping 30s")
                    await asyncio.sleep(30)
                    continue

                for sym in symbols:
                    await ws.send(json.dumps({"type": "subscribe", "symbol": sym}))

                while True:
                    msg = await asyncio.wait_for(ws.recv(), timeout=30)
                    data = json.loads(msg)
                    if data.get("type") == "trade":
                        for trade in data.get("data", []):
                            await flush_trade(redis, trade)

        except asyncio.TimeoutError:
            logger.warning("[WS] Timeout — retrying")
        except websockets.ConnectionClosed:
            logger.warning("[WS] Connection closed — reconnecting")
        except Exception as e:
            logger.error(f"[WS] Error: {e}")

        logger.info(f"[WS] Reconnecting in {reconnect_delay}s...")
        await asyncio.sleep(reconnect_delay)
        reconnect_delay = min(reconnect_delay * 2, max_delay)

# --- Boot ---
if __name__ == "__main__":
    try:
        logger.info("🚀 Starting real-time streamer with OHLCV updates every 10s")
        asyncio.run(stream_trades())
    except Exception as e:
        logger.exception(f"[FATAL] Crashed: {e}")

File: data/src/bin/websocket.rs
use std::{collections::HashMap, env, time::Duration};

use chrono::{Utc, TimeZone};
use dotenv::dotenv;
use futures::{stream::StreamExt, SinkExt};
use redis::AsyncCommands;
use serde::{Deserialize, Serialize};
use tokio::{task::LocalSet, time::sleep};
use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};

const SYMBOLS_KEY: &str = "stock:symbols";
const PRICE_PREFIX: &str = "stock:price:";
const TRADE_PREFIX: &str = "stock:trade:";
const OHLCV_PREFIX: &str = "stock:ohlcv:";

#[derive(Debug, Deserialize)]
struct WebSocketMessage {
    r#type: String,
    data: Option<Vec<TradeData>>,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
struct TradeData {
    s: String,     // symbol
    p: f64,        // price
    v: Option<f64>,// volume
    t: i64,        // trade time in ms since epoch
}

#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv().ok();
    let local = LocalSet::new();

    local
        .run_until(async {
            if let Err(e) = run().await {
                eprintln!("❌ Application error: {}", e);
            }
        })
        .await;

    Ok(())
}

async fn run() -> Result<(), Box<dyn std::error::Error>> {
    let api_key = env::var("FINNHUB_API_KEY")?;
    let redis_url = env::var("REDIS_URL")?;

    // --- Auto-handle TLS for Redis ---
    let redis_client = redis::Client::open(redis_url.clone())?;
    if redis_url.starts_with("rediss://") {
        println!("🔐 Connecting to Redis with TLS...");
    } else {
        println!("🌐 Connecting to Redis without TLS...");
    }

    let mut redis = redis_client.get_multiplexed_async_connection().await?;
    println!("✅ Connected to Redis");

    // WebSocket URL
    let ws_url = url::Url::parse(&format!("wss://ws.finnhub.io?token={}", api_key))?;

    // OHLCV in-memory state: symbol -> (open, high, low, close, volume)
    let mut ohlcv_map: HashMap<String, (f64, f64, f64, f64, f64)> = HashMap::new();

    let mut reconnect_delay = Duration::from_secs(3);

    loop {
        println!("🌐 Attempting connection to Finnhub WebSocket...");

        match connect_async(ws_url.clone()).await {
            Ok((mut ws_stream, _)) => {
                println!("✅ WebSocket connected successfully.");
                reconnect_delay = Duration::from_secs(3);
                let mut last_symbols = Vec::new();

                loop {
                    // Refresh subscriptions
                    match redis.smembers::<_, Vec<String>>(SYMBOLS_KEY).await {
                        Ok(current_symbols) => {
                            if current_symbols != last_symbols {
                                last_symbols = current_symbols.clone();

                                if current_symbols.is_empty() {
                                    println!("⚠️ No stock symbols in '{}'", SYMBOLS_KEY);
                                    continue;
                                }

                                println!(
                                    "🔄 Updating subscriptions for {} symbols...",
                                    current_symbols.len()
                                );
                                for sym in &current_symbols {
                                    let msg =
                                        format!(r#"{{"type":"subscribe","symbol":"{}"}}"#, sym);
                                    if let Err(e) = ws_stream.send(Message::Text(msg.into())).await {
                                        eprintln!("❌ Failed to subscribe {}: {}", sym, e);
                                    }
                                    sleep(Duration::from_millis(50)).await;
                                }
                            }
                        }
                        Err(e) => {
                            eprintln!("❌ Redis symbol fetch error: {}", e);
                            sleep(Duration::from_secs(5)).await;
                            continue;
                        }
                    }

                    // Process incoming WebSocket messages
                    while let Some(msg) = ws_stream.next().await {
                        match msg {
                            Ok(Message::Text(text)) => {
                                if let Ok(parsed) =
                                    serde_json::from_str::<WebSocketMessage>(&text)
                                {
                                    if parsed.r#type == "trade" {
                                        if let Some(trades) = parsed.data {
                                            let mut redis_conn = match redis_client
                                                .get_multiplexed_async_connection()
                                                .await
                                            {
                                                Ok(conn) => conn,
                                                Err(e) => {
                                                    eprintln!(
                                                        "❌ Redis reconnect error: {}",
                                                        e
                                                    );
                                                    continue;
                                                }
                                            };

                                            for trade in trades {
                                                let symbol = trade.s.clone();
                                                let price = trade.p;
                                                let volume = trade.v.unwrap_or(0.0);

                                                // Convert Finnhub's trade.t (ms since epoch) to RFC3339
                                                let trade_time = Utc.timestamp_millis_opt(trade.t)
                                                    .single()
                                                    .expect("Invalid trade timestamp");
                                                let trade_time_str = trade_time.to_rfc3339();

                                                // 1) Price + trade info
                                                let _ = redis_conn
                                                    .set::<_, _, ()>(
                                                        format!("{}{}", PRICE_PREFIX, symbol),
                                                        price,
                                                    )
                                                    .await;
                                                let _ = redis_conn
                                                    .hset_multiple::<_, _, _, ()>(
                                                        format!("{}{}", TRADE_PREFIX, symbol),
                                                        &[
                                                            ("price".to_string(),
                                                                price.to_string()),
                                                            ("timestamp".to_string(),
                                                                trade.t.to_string()),
                                                            ("volume".to_string(),
                                                                volume.to_string()),
                                                            ("updated_at".to_string(),
                                                                trade_time_str.clone()),
                                                        ],
                                                    )
                                                    .await;

                                                // 2) Update OHLCV state
                                                let entry = ohlcv_map
                                                    .entry(symbol.clone())
                                                    .or_insert((
                                                        price, // open
                                                        price, // high
                                                        price, // low
                                                        price, // close
                                                        0.0,   // volume
                                                    ));
                                                entry.1 = entry.1.max(price); // high
                                                entry.2 = entry.2.min(price); // low
                                                entry.3 = price; // close
                                                entry.4 += volume; // volume

                                                // 3) Immediate OHLCV flush to Redis
                                                let fields = [
                                                    ("open".to_string(), entry.0.to_string()),
                                                    ("high".to_string(), entry.1.to_string()),
                                                    ("low".to_string(), entry.2.to_string()),
                                                    ("close".to_string(), entry.3.to_string()),
                                                    ("volume".to_string(), entry.4.to_string()),
                                                    ("updated_at".to_string(),
                                                        trade_time_str.clone()),
                                                ];
                                                let _ = redis_conn
                                                    .hset_multiple::<_, _, _, ()>(
                                                        format!("{}{}", OHLCV_PREFIX, symbol),
                                                        &fields,
                                                    )
                                                    .await;
                                            }
                                        }
                                    }
                                }
                            }
                            Ok(_) => {}
                            Err(e) => {
                                eprintln!("❌ WebSocket stream error: {}", e);
                                break;
                            }
                        }
                    }

                    println!("🔁 WebSocket disconnected. Retrying...");
                    break;
                }
            }
            Err(e) => {
                eprintln!("❌ Connection error: {}", e);
            }
        }

        println!("⏳ Waiting {}s before retry...", reconnect_delay.as_secs());
        sleep(reconnect_delay).await;
        reconnect_delay = (reconnect_delay * 2).min(Duration::from_secs(60));
    }
}
